File: backend/schemas/oauth.py
"""Pydantic-Schemas fuer OAuth-Provider-Konfiguration.

Sicherheits-Invarianten:
- Response-Modelle geben Client-Secrets NUR maskiert zurueck (max. 4 Zeichen Sicht).
- Create/Update akzeptieren Client-Secrets im Klartext und speichern sie
  Fernet-encrypted (im Service-Layer, nicht hier).
- ``client_secret`` ist im Create-Body OPTIONAL — einige Provider (z. B. PKCE-only)
  benoetigen keins.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

# ── Anzeige-Presets (fuer UI-Dropdowns) ────────────────────────────────

OAUTH_PRESETS: tuple[str, ...] = (
    "google",
    "discord",
    "github",
    "microsoft",
    "twitter",
    "custom_oidc",
    "custom_oauth2",
)


# ── Create ─────────────────────────────────────────────────────────────

class OAuthProviderCreate(BaseModel):
    slug: str = Field(min_length=2, max_length=64, pattern=r"^[a-z0-9][a-z0-9_-]*$")
    name: str = Field(min_length=1, max_length=128)
    preset: str
    enabled: bool = True
    client_id: str = Field(min_length=1, max_length=512)
    # Optional: leeres/fehlendes Secret bedeutet "kein Secret hinterlegt".
    client_secret: str | None = Field(default=None, max_length=4096)

    # Optionale Overrides fuer custom_*-Presets
    issuer: str | None = Field(default=None, max_length=512)
    authorization_endpoint: str | None = Field(default=None, max_length=512)
    token_endpoint: str | None = Field(default=None, max_length=512)
    userinfo_endpoint: str | None = Field(default=None, max_length=512)
    scope: str | None = Field(default=None, max_length=512)
    claims_mapping_json: str | None = None
    position: int = 0


# ── Update ─────────────────────────────────────────────────────────────

class OAuthProviderUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    enabled: bool | None = None
    client_id: str | None = Field(default=None, min_length=1, max_length=512)
    # Falls None/"" → Secret bleibt unveraendert. "" wird als "Secret loeschen" interpretiert.
    # Frontend MUSS maskierten Wert (****1234) zurueckschicken, wenn das Secret
    # erhalten bleiben soll — der Service prueft das Praefix und ueberspringt.
    client_secret: str | None = Field(default=None, max_length=4096)
    issuer: str | None = Field(default=None, max_length=512)
    authorization_endpoint: str | None = Field(default=None, max_length=512)
    token_endpoint: str | None = Field(default=None, max_length=512)
    userinfo_endpoint: str | None = Field(default=None, max_length=512)
    scope: str | None = Field(default=None, max_length=512)
    claims_mapping_json: str | None = None
    position: int | None = None


# ── Response ───────────────────────────────────────────────────────────

class OAuthProviderResponse(BaseModel):
    id: int
    slug: str
    name: str
    preset: str
    enabled: bool
    client_id: str
    # Maskiert, s. _mask_secret
    client_secret: str
    issuer: str | None = None
    authorization_endpoint: str | None = None
    token_endpoint: str | None = None
    userinfo_endpoint: str | None = None
    scope: str | None = None
    claims_mapping_json: str | None = None
    position: int
    created_at: str
    updated_at: str

    model_config = ConfigDict(from_attributes=True)


class OAuthProviderPublic(BaseModel):
    """Schlanke Public-Variante fuer das Login-UI (keine Client-IDs noetig)."""

    slug: str
    name: str
    preset: str
    position: int


# ── Test-Verbindung ────────────────────────────────────────────────────

class OAuthTestResult(BaseModel):
    ok: bool
    message: str

File: frontend/src/api/oauth.ts
import { api } from '@/api/client'

export const OAUTH_PRESETS = [
  'google',
  'discord',
  'github',
  'microsoft',
  'twitter',
  'custom_oidc',
  'custom_oauth2',
] as const

export type OAuthPreset = (typeof OAUTH_PRESETS)[number]

export interface OAuthProvider {
  id: number
  slug: string
  name: string
  preset: string
  enabled: boolean
  client_id: string
  client_secret: string
  issuer: string | null
  authorization_endpoint: string | null
  token_endpoint: string | null
  userinfo_endpoint: string | null
  scope: string | null
  claims_mapping_json: string | null
  position: number
  redirect_uri?: string
  created_at: string
  updated_at: string
}

export interface OAuthProviderPublic {
  slug: string
  name: string
  preset: string
  position: number
}

export interface OAuthSwitches {
  allow_registration: boolean
  allow_linking: boolean
  require_verified_email: boolean
}

export interface OAuthTestResult {
  ok: boolean
  message: string
}

export interface OAuthUserLink {
  id: number
  provider_id: number
  provider_slug: string
  provider_name: string
  provider_preset: string
  created_at: string
  last_used_at: string | null
}

export interface OAuthProviderCreate {
  slug: string
  name: string
  preset: string
  enabled?: boolean
  client_id: string
  client_secret?: string | null
  issuer?: string | null
  authorization_endpoint?: string | null
  token_endpoint?: string | null
  userinfo_endpoint?: string | null
  scope?: string | null
  claims_mapping_json?: string | null
  position?: number
}

export type OAuthProviderUpdate = Partial<OAuthProviderCreate>

export const oauthApi = {
  listPublicProviders: () =>
    api<OAuthProviderPublic[]>('/oauth/public/providers'),

  listProviders: () => api<OAuthProvider[]>('/oauth/providers'),
  getProvider: (id: number) => api<OAuthProvider>(`/oauth/providers/${id}`),
  createProvider: (body: OAuthProviderCreate) =>
    api<OAuthProvider>('/oauth/providers', { method: 'POST', body: JSON.stringify(body) }),
  updateProvider: (id: number, body: OAuthProviderUpdate) =>
    api<OAuthProvider>(`/oauth/providers/${id}`, { method: 'PATCH', body: JSON.stringify(body) }),
  deleteProvider: (id: number) =>
    api<{ message: string }>(`/oauth/providers/${id}`, { method: 'DELETE' }),
  updateSecret: (id: number, client_secret: string) =>
    api<{ message: string }>(`/oauth/providers/${id}/secret`, {
      method: 'POST',
      body: JSON.stringify({ client_secret }),
    }),
  testProvider: (id: number) =>
    api<OAuthTestResult>(`/oauth/providers/${id}/test`, { method: 'POST' }),

  getSwitches: () => api<OAuthSwitches>('/oauth/switches'),
  updateSwitches: (body: Partial<OAuthSwitches>) =>
    api<OAuthSwitches>('/oauth/switches', { method: 'PATCH', body: JSON.stringify(body) }),

  listMyLinks: () => api<OAuthUserLink[]>('/oauth/me/links'),
  unlinkProvider: (providerId: number) =>
    api<{ message: string }>(`/oauth/me/links/${providerId}`, { method: 'DELETE' }),
}
